src/main/routing/packet.c

```diff
@@ -128,10 +130,27 @@ void packet_setPayload(Packet* packet, const Thread* thread, PluginVirtualPtr pa
 
     /* the payload starts with 1 ref, which we hold */
     packet->payload = payload_new(thread, payload, payloadLength);
+    utility_alwaysAssert(packet->payload != NULL);
     /* application data needs a priority ordering for FIFO onto the wire */
     packet->priority = host_getNextPacketPriority(thread_getHost(thread));
 }
 
+/* This is a copy of `packet_setPayload` but passes the memory manager through. Once we've moved UDP
+ * sockets to rust, we can remove `packet_setPayload` and rename this function to
+ * `packet_setPayload`. */
+void packet_setPayloadWithMemoryManager(Packet* packet, const Host* host, PluginVirtualPtr payload,
+                                        gsize payloadLength, MemoryManager* mem) {
```

Comment:
I think `MemoryManager*` can be `const` here and the other functions that are only reading memory?

I'm also a little nervous about handing mutable pointers to Rust objects to C code in general, since it's a bit difficult to be confident that we won't break Rust aliasing rules. Instead of passing the `MemoryManager` around maybe we could pass the `const ProcessRefCell*` around and use the existing `process_readPtr` and `process_writePtr` APIs?

Reply:
> We call into the C code from LegacyTcpSocket::recvmsg(), which takes a &mut MemoryManager as an argument, so using process_writePtr would cause a double mutable borrow. 

Ah I see.

> Aside from signals and multi-threading, there wouldn't be any way to break aliasing rules.

I think another issue is Rust not enforcing borrowing rules when a pointer is propagated through the callstack. So e.g. we might have:

```
fn rust_fn(mem: &mut MemoryManager, ptr: TypedPluginPtr<u8>) {
  let first_borrow = mem.writer(ptr);
  unsafe { c_fn(mem as *mut _)};
```

and

```
void c_fn(MemoryManager* mem) {
  memorymanager_writePtr(mem, ...);
}
```

then `memorymanager_writePtr`  would construct a `&mut MemoryManager` when it's already borrowed further up the call stack.

I suppose the root cause of this bug is passing a mutable pointer to `MemoryManager` to `c_fn` inside an `unsafe` block while there is already a live reference to `MemoryManager`, knowing that the pointer may be used to create another, incompatible, reference. Maybe we can be careful to avoid such situations at those points.

Having worked through it, I think I'm leaning towards how you currently have it, but up to you.

Reply:
If we stick with the current approach, maybe the new `memorymanager_` functions should be marked `unsafe`, with a `safety` comment that there must not exist (incompatible) borrows/references to the pointers being passed?

I know I haven't been super rigorous on this in general, and have been leaning on trying to only pass `const` pointers, though I suppose a similar condition exists even then that the caller must not hold a mutable borrow/reference.

Reply:
> If we stick with the current approach, maybe the new memorymanager_ functions should be marked unsafe, with a safety comment that there must not exist (incompatible) borrows/references to the pointers being passed?
>
> I know I haven't been super rigorous on this in general, and have been leaning on trying to only pass const pointers, though I suppose a similar condition exists even then that the caller must not hold a mutable borrow/reference.

Err, I guess that doesn't directly address the example I gave either; the safety comment would belong on `c_fn`, which would be e.g. `tcp_sendUserData`. It's a little weird to have Rust safety comments on C code; I guess this ispart of the implicit safety conditions when calling a foreign C function with pointers to Rust objects...

Reply:
Maybe it's not so bad - the conversion to a pointer itself requires borrowing the corresponding reference to the original object. It's possible to mess this up, but not trivially. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=09dec6cb6b490c7ac04b7cdb38ed81b6

```
fn takes_ptr(p: *mut i32) {
    unsafe { *p += 1}
}

pub fn main() {
    let mut x = 42;
    
    /* borrow checker prevents this
    let borrow = &mut x;
    takes_ptr(&mut x as *mut _);
    println!("{borrow}");
    println!("{x}");
    */
    
    // Doesn't catch this unsoundness, but at this point we're really going out of our way
    // to fool the checker
    let ptr = &mut x as *mut _;
    let borrow = &mut x;
    takes_ptr(ptr);
    println!("{borrow}");
    println!("{x}");
}
```

Reply:
> Maybe it's not so bad - the conversion to a pointer itself requires borrowing the corresponding reference to the original object. It's possible to mess this up, but not trivially. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=09dec6cb6b490c7ac04b7cdb38ed81b6

Yeah, I think if you only convert a reference to a pointer at the C function callsite then you shouldn't run into issues (unless the C function itself does something that would break aliasing or lifetime rules). You don't need to even explicitly cast to a pointer in this case since rust will do it implicitly for you (for example `let borrow = &mut x; takes_ptr(borrow);`), maybe because it's always safe to do?

> Having worked through it, I think I'm leaning towards how you currently have it, but up to you.

Okay I think I'll leave it as-is, but I'll add some comments.